### PR TITLE
feat(trie): trie cursor abstraction

### PR DESCRIPTION
**Pull request by rkrasiuk**
## Description

Create an abstraction for trie cursor that would allow using state root calculator independently of the database.

### CODE REVIEW

File: crates/trie/src/trie_cursor/noop.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        &self,
+        _hashed_address: reth_primitives::B256,
+    ) -> Result<Box<dyn TrieCursor<Key = StoredNibblesSubKey> + '_>, DatabaseError> {
+        Ok(Box::new(NoopStorageTrieCursor::default()))
+    }
+}
+
+/// Noop account trie cursor.
+#[derive(Default, Debug)]
+pub struct NoopAccountTrieCursor;
```
Comment (DaniPopes) [inline crates/trie/src/trie_cursor/noop.rs:29]: `#[non_exhaustive]`?

File: crates/trie/src/trie.rs [hunk 2]
```diff
@@ -80,12 +80,9 @@ impl<'a, TX, H> StateRoot<'a, TX, H> {
     }
 
     /// Set the hashed cursor factory.
-    pub fn with_hashed_cursor_factory<HF>(
-        self,
-        hashed_cursor_factory: HF,
-    ) -> StateRoot<'a, TX, HF> {
+    pub fn with_hashed_cursor_factory<HF>(self, hashed_cursor_factory: HF) -> StateRoot<T, HF> {
```
Comment (dvush) [inline crates/trie/src/trie.rs:83]: it would be nice to have `with_trie_cursor_factory` too

Comment (rkrasiuk) [reply crates/trie/src/trie.rs:83]: good point, added

**Comment by onbjerg** [comment]
i guess this needs to go in the new alloy-trie crate?

**Comment by DaniPopes** [comment]
> i guess this needs to go in the new alloy-trie crate?

No, this doesn't touch code in primitives

**Review by mattsse — APPROVED** [review]
these seem reasonable.

pending @DaniPopes

**Review by DaniPopes — APPROVED** [review]

**Pull request merged**